### PR TITLE
Detect circular dependencies during the build process in Bob

### LINKED ISSUE #3730: Engine crashes with no error message when doing circular require (DEF-1578)
Engine crashes with no error message when doing circular require in Lua.

It would be nice if it posted some kind of error message to the console or similiar when this happens. Atleast in debug builds when running the game from the editor. Otherwise it is really hard to debug what really happened.

### PR DESCRIPTION
**Pull request by AGulev**
Before this fix, it was possible to bundle a project with circular dependencies, which would result in a runtime crash. Now, Bob detects such cases and reports them during the build process:

```
Circular dependency detected:
main.lua class com.dynamo.bob.pipeline.ScriptBuilders$LuaScriptBuilder
-> first.lua class com.dynamo.bob.pipeline.ScriptBuilders$LuaScriptBuilder (Circular Point)
second.lua class com.dynamo.bob.pipeline.ScriptBuilders$LuaScriptBuilder
third.lua class com.dynamo.bob.pipeline.ScriptBuilders$LuaScriptBuilder
-> first.lua class com.dynamo.bob.pipeline.ScriptBuilders$LuaScriptBuilder (Circular Point)'
```

Fix https://github.com/defold/defold/issues/3730

## PR checklist

* [x] Code
	* [x] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/LuaBuilderTest.java [hunk 1]
```diff
@@ -29,6 +25,8 @@
 import com.dynamo.properties.proto.PropertiesProto.PropertyDeclarations;
 import com.dynamo.properties.proto.PropertiesProto.PropertyDeclarationEntry;
 
+import static org.junit.Assert.*;
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/LuaBuilderTest.java:28]: 😱

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/LuaBuilderTest.java [hunk 2]
```diff
@@ -287,4 +283,18 @@ public void testLuaBytecodeDeltaCalculation() throws Exception {
         // byte - the last diffing byte
         assertTrue(delta[261] == 99);
     }
+
+    @Test
+    public void testCircularRequire() throws Exception {
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/LuaBuilderTest.java:288]: 👍

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java [hunk 3]
```diff
@@ -478,16 +468,20 @@ public Task<?> createTask(IResource inputResource) throws CompileExceptionError
      * @return task
      * @throws CompileExceptionError
      */
-    public Task<?> createTask(IResource inputResource, Class<? extends Builder<?>> builderClass) throws CompileExceptionError {
+    public Task createTask(IResource inputResource, Class<? extends Builder> builderClass) throws CompileExceptionError {
         // It's possible to build the same resource using different builders
         String key = inputResource.getPath()+" "+builderClass;
-        Task<?> task = tasks.get(key);
+        if (!circularDependencyChecker.add(key)) {
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java:484]: A very neat solution! I like it.

Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java:484]: This solution became possible after refactoring and building the dependency tree during the creation stage. It made the implementation of such a detector really trivial.

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/LuaBuilderTest.java (hunk 1 above)
Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/LuaBuilderTest.java]: fixed

**Review by britzl — APPROVED** [review]

**Pull request merged**